### PR TITLE
Implement pointerMove webdriver action

### PR DESCRIPTION
**Pull request by georgeroman**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23996)

### CODE REVIEW

File: components/compositing/compositor.rs [hunk 1]
```diff
@@ -522,19 +522,28 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
...
             ) => {
+                let dppx = self.device_pixels_per_page_px();
+                let point = dppx.transform_point(Point2D::new(x, y));
                 self.on_mouse_window_event_class(match mouse_event_type {
                     MouseEventType::Click => {
-                        MouseWindowEvent::Click(mouse_button, DevicePoint::new(x, y))
-                    },
-                    MouseEventType::MouseDown => {
-                        MouseWindowEvent::MouseDown(mouse_button, DevicePoint::new(x, y))
+                        MouseWindowEvent::Click(mouse_button, DevicePoint::new(point.x, point.y))
```
Comment (jdm) [inline components/compositing/compositor.rs]: dppx.transform_point returns a Point2D<f32, DevicePixel>, which is the same as DevicePoint. We should be able to pass `point` directly instead of calling `DevicePoint::new` in all of these branches.

File: components/script/webdriver_handlers.rs [hunk 2]
```diff
@@ -270,6 +272,49 @@ pub fn handle_get_browsing_context_id(
     reply.send(result).unwrap()
 }
 
+fn get_element_in_view_center_point(element: &Element) -> Option<Point2D<i64>> {
+    match element.GetClientRects().iter().next() {
+        Some(rectangle) => {
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: This can be a `map` call instead of a match.

File: components/script/webdriver_handlers.rs [hunk 3]
```diff
@@ -270,6 +272,49 @@ pub fn handle_get_browsing_context_id(
     reply.send(result).unwrap()
 }
 
+fn get_element_in_view_center_point(element: &Element) -> Option<Point2D<i64>> {
+    match element.GetClientRects().iter().next() {
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: Why are we getting all the rects and only using the first one?

File: components/script/webdriver_handlers.rs [hunk 4]
```diff
@@ -270,6 +272,49 @@ pub fn handle_get_browsing_context_id(
...
+        },
+        _ => None,
+    }
+}
+
+pub fn handle_get_element_in_view_center_point(
+    documents: &Documents,
+    pipeline: PipelineId,
+    element_id: String,
+    reply: IpcSender<Result<Option<(i64, i64)>, ()>>,
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: This will need updating with #23951

File: components/webdriver_server/actions.rs [hunk 5]
```diff
@@ -156,22 +169,32 @@ impl Handler {
...
+        Ok(())
     }
 
     // https://w3c.github.io/webdriver/#dfn-dispatch-a-keydown-action
-    fn dispatch_keydown_action(&mut self, source_id: &str, action: &KeyDownAction) {
+    fn dispatch_keydown_action(
+        &mut self,
+        source_id: &str,
+        action: &KeyDownAction,
+    ) -> Result<(), ErrorStatus> {
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: Why this change if no errors can occur? Same question for the other dispatch methods that were changed in the same way.

File: components/webdriver_server/actions.rs [hunk 6]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+        {
+            InputSourceState::Null => unreachable!(),
+            InputSourceState::Key(_) => unreachable!(),
+            InputSourceState::Pointer(pointer_input_state) => {
+                (pointer_input_state.x as i64, pointer_input_state.y as i64)
+            },
+        };
+
+        let x_offset = action.x.unwrap_or(0);
+        let y_offset = action.y.unwrap_or(0);
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: When can the x and y values be unspecified?

File: components/webdriver_server/actions.rs [hunk 7]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+                let (sender, receiver) = ipc::channel().unwrap();
+                self.top_level_script_command(WebDriverScriptCommand::GetElementInViewCenterPoint(
+                    x.to_string(),
+                    sender,
+                ))
+                .unwrap();
+
+                match receiver.recv().unwrap() {
+                    Ok(point) => match point {
+                        Some((x_element, y_element)) => (x_element, y_element),
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: This can be written as `Some(point) => point,`

File: components/webdriver_server/actions.rs [hunk 8]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+        if x < 0 || x as f32 > viewport.width || y < 0 || y as f32 > viewport.height {
+            return Err(ErrorStatus::MoveTargetOutOfBounds);
+        }
+
+        let duration = match action.duration {
+            Some(duration) => duration,
+            None => tick_duration,
+        };
+
+        thread::sleep(Duration::from_millis(17));
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: What is the purpose of this sleep?

File: components/webdriver_server/actions.rs [hunk 9]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+                .unwrap();
+        }
+
+        if last {
+            return Ok(());
+        }
+
+        thread::sleep(Duration::from_millis(17));
+
+        self.perform_pointer_move(
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: Rather than writing this recursively, can we use a loop?

File: components/webdriver_server/actions.rs [hunk 10]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+    }
+
+    // https://w3c.github.io/webdriver/#dfn-dispatch-a-pointermove-action
+    fn dispatch_pointermove_action(
+        &mut self,
+        source_id: &str,
+        action: &PointerMoveAction,
+        tick_duration: u64,
+    ) -> Result<(), ErrorStatus> {
+        let tick_start = Instant::now();
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: Could we add comments with step numbers to this method?

File: components/webdriver_server/actions.rs [hunk 11]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+        };
+
+        thread::sleep(Duration::from_millis(17));
+
+        self.perform_pointer_move(source_id, duration, start_x, start_y, x, y, &tick_start)?;
+
+        Ok(())
+    }
+
+    fn perform_pointer_move(
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: Add a link to https://w3c.github.io/webdriver/#dfn-perform-a-pointer-move

File: components/webdriver_server/actions.rs [hunk 12]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+        &mut self,
+        source_id: &str,
+        duration: u64,
+        start_x: i64,
+        start_y: i64,
+        target_x: i64,
+        target_y: i64,
+        tick_start: &Instant,
+    ) -> Result<(), ErrorStatus> {
+        let time_delta = tick_start.elapsed().as_millis();
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: Can we add comments with step numbers?

File: components/webdriver_server/actions.rs [hunk 13]
```diff
@@ -317,5 +358,148 @@ impl Handler {
...
+            self.constellation_chan
+                .send(ConstellationMsg::WebDriverCommand(cmd_msg))
+                .unwrap();
+        }
+
+        if last {
+            return Ok(());
+        }
+
+        thread::sleep(Duration::from_millis(17));
```
Comment (jdm) [inline components/webdriver_server/actions.rs]: The current implementation is synchronous, so we don't match the constraints described by https://w3c.github.io/webdriver/#issue-container-generatedID-46. I think we should address that spawning a thread that is responsible for sending the remaining webdriver commands.

File: components/script/webdriver_handlers.rs [hunk 14]
```diff
@@ -270,6 +272,49 @@ pub fn handle_get_browsing_context_id(
...
+            let height = rectangle.Height().round() as i64;
+
+            let window = window_from_node(element.upcast::<Node>());
+            let innerWidth = window.InnerWidth() as i64;
+            let innerHeight = window.InnerHeight() as i64;
+
+            let left = cmp::max(0, cmp::min(x, x + width));
+            let right = cmp::min(innerWidth, cmp::max(x, x + width));
+            let top = cmp::max(0, cmp::min(y, y + height));
+            let bottom = cmp::min(innerHeight, cmp::max(y, y + height));
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: It's not clear to me that comparing against innerWidth and innerHeight will work when an element is outside the initial viewport and the page is scrolled so that the element is in view.

File: components/script/webdriver_handlers.rs [hunk 15]
```diff
@@ -270,6 +272,49 @@ pub fn handle_get_browsing_context_id(
...
+            let x = (left + right).checked_div(2).unwrap();
+            let y = (top + bottom).checked_div(2).unwrap();
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: Why are we using checked_div and then unwrapping the result?

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/webdriver_handlers.rs (hunk 3 above)
Comment (georgeroman) [reply components/script/webdriver_handlers.rs]: That's the behavior from the spec: https://w3c.github.io/webdriver/#dfn-center-point.

File: components/webdriver_server/actions.rs (hunk 6 above)
Comment (georgeroman) [reply components/webdriver_server/actions.rs]: Steps 9 and 12 from https://w3c.github.io/webdriver/#dfn-process-a-pointer-move-action accept undefined values for `x` and `y` though their default value is not specified anywhere.

File: components/webdriver_server/actions.rs (hunk 8 above)
Comment (georgeroman) [reply components/webdriver_server/actions.rs]: This is step 9.1 from https://w3c.github.io/webdriver/#dfn-process-a-pointer-move-action. The actual value of `17` milliseconds is taken from https://searchfox.org/mozilla-central/source/testing/marionette/action.js#1377. I'll add a comment describing its meaning.

File: components/script/webdriver_handlers.rs (hunk 3 above)
Comment (jdm) [reply components/script/webdriver_handlers.rs]: Let's add a spec link and some comments with step numbers, in that case :)

File: components/webdriver_server/actions.rs (hunk 13 above)
Comment (georgeroman) [reply components/webdriver_server/actions.rs]: I'm not sure how I could do this while also allowing for errors triggered while performing the move to be propagated back to the main thread.

File: components/script/webdriver_handlers.rs (hunk 14 above)
Comment (georgeroman) [reply components/script/webdriver_handlers.rs]: This will not work in that case, it needs to take into account the whole available viewport not just the window. I'll address this issue. However, this case won't come up as scrolling into view is not supported yet.

File: components/webdriver_server/actions.rs (hunk 13 above)
Comment (georgeroman) [reply components/webdriver_server/actions.rs]: Since `perform_move_action` needs access to `self`, we'll also need to synchronize its shared usage, is that right?

File: components/script/webdriver_handlers.rs (hunk 14 above)
Comment (georgeroman) [reply components/script/webdriver_handlers.rs]: Actually, I think this works. `innerWidth` and `innerHeight` should reflect the size of the whole viewport, not just the size of the window, if I understood correctly.

Comment (jdm) [reply components/script/webdriver_handlers.rs]: The viewport is the window size.

File: components/webdriver_server/actions.rs (hunk 13 above)
Comment (jdm) [reply components/webdriver_server/actions.rs]: I recommend making perform_move_action a free function that accepts all of the state it requires as arguments. Can the input_state_table entry for a particular source id change? If not, we can pass the pointer input state and the constellation channel. If it can, then we can store the input state table inside of an Arc<Mutex> and pass that along with the constellation channel.

**Review by jdm — CHANGES_REQUESTED** [review]
With the change to make subsequent pointer movements happen asynchronously, this should be ready to merge.

**Comment by georgeroman** [comment]
Done.

File: components/webdriver_server/actions.rs (hunk 13 above)
Comment (georgeroman) [reply components/webdriver_server/actions.rs]: Running `perform_move_action` in a separate thread interferes with the `ElementClick` command, where we need to dispatch a `pointerMove`, a `pointerDown` and then a `pointerUp` action, in the proper order: https://w3c.github.io/webdriver/#element-click (steps 8.10 - 8.13). The `pointerMove` action might happen after the `pointerDown` and `pointerUp` actions, which is not desirable.
